cas-server-core/src/main/java/org/jasig/cas/CentralAuthenticationServiceImpl.java

```diff
@@ -284,11 +285,21 @@ public ServiceTicket grantServiceTicket(
                     uniqueTicketIdGenKey, service.getId());
             throw new UnauthorizedSsoServiceException();
         }
-        
+
+        final List<Authentication> authentications = ticketGrantingTicket.getChainedAuthentications();
+        final Principal principal = authentications.get(authentications.size() - 1).getPrincipal();
+
+        final Map<String, Object> principalAttrs = registeredService.getAttributeReleasePolicy().getAttributes(principal);
+        if (!registeredService.getAuthorizationStrategy().isServiceAccessAuthorizedForPrincipal(principalAttrs, service)) {
+            logger.warn("ServiceManagement: Cannot grant service ticket because Service [{}] is not authorized for use by [{}].",
+                    service.getId(), principal);
+            throw new UnauthorizedServiceForPrincipalException();

```

Comment:
is this going to be enough information for a client to construct a meaningful message (since I am assuming you won't be redirected back to the failing service?  I.e. the ability to display a custom failure message PER service ("If you feel you should have access to this service, please call (XXX) XXX-XXXX) or visit here"